### PR TITLE
fadvise: new command wrapping posix_fadvise

### PR DESCRIPTION
**Pull request by masatake**
For measuring and monitoring the performance aspect of a command, people may want to drop the page caches related to specified files.

In some cases, writing 1 to /proc/sys/vm/drop_caches is overkill.  The fadvise command can be used for dropping page caches related to specified files; the impact on a system is much smaller than /proc/sys/vm/drop_caches.

Signed-off-by: Masatake YAMATO <yamato@redhat.com>

### CODE REVIEW

**Comment by masatake** [comment]
I misunderstood the word "advice" is countable. I will fix the man page and commit log.

I wonder if advice other than "dontneed" is useful in command level or not.
If there is no use case, I will rename the command to fdrop or fdontneed. Comments are welcome.

File: misc-utils/fadvise.c [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+			if (args.fd < 0)
+				errx(EXIT_FAILURE, _("invalid file descriptor value specified"));
+			break;
+		case 'l':
+			args.len = cvtnum(optarg);
+			if (args.len < 0)
+				errx(EXIT_FAILURE, _("invalid length value specified"));
+			break;
+		case 'o':
+			args.offset = cvtnum(optarg);
```
Comment (karelzak) [inline misc-utils/fadvise.c:138]: What about to use  
```
   args.offset = strtosize_or_err(optarg, _("invalid offset value specified"));
```

File: misc-utils/fadvise.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+	if (argc - optind > 1) {
+		warnx(_("specify one file descriptor or file name"));
+		errtryhelp(EXIT_FAILURE);
+	}
+
+	if (args.fd == -1) {
+		args.fd = open(argv[optind], O_RDONLY);
+		if (args.fd < 0)
+			err(EXIT_FAILURE, _("failed to open: %s"), argv[optind]);
```
Comment (karelzak) [inline misc-utils/fadvise.c:169]: We usually use "cannot open %s"   (see po/util-linux.pot for more details).

File: misc-utils/fadvise.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		do_close = true;
+	}
+
+	rc = fadvise_args(&args);
```
Comment (karelzak) [inline misc-utils/fadvise.c:173]: I don't think we need to encapsulate the call in fadvise_args(); just call posix_fadvise() in main().

File: misc-utils/fadvise.c (hunk 1 above)
Comment (masatake) [reply misc-utils/fadvise.c]: I replaced all `cvtnum` with `strto*_or_err` functions.

File: misc-utils/fadvise.c (hunk 2 above)
Comment (masatake) [reply misc-utils/fadvise.c]: Updated.

File: misc-utils/fadvise.c (hunk 3 above)
Comment (masatake) [reply misc-utils/fadvise.c]: Updated. `args` is also redundant. So I deleted the structure.

**Comment by karelzak** [comment]
> I wonder if advice other than "dontneed" is useful in command level or not. If there is no use case, I will rename the command to fdrop or fdontneed. Comments are welcome.

I think "dontneed" seems fine; let's make available on the command line all the features; it's up to users to get any use-case ;-)

**Comment by neheb** [comment]
```
meson.build:151:1: warning unused variable fadvise_sources
151 | fadvise_sources = files(
```

Does not get built with meson.

**Comment by masatake** [comment]
It will be nice if we can detect such a build failure before merging.
Isn't it possible to integrate "building wtth meason" with our CI/CD environment?

**Comment by neheb** [comment]
that output is from https://sr.ht/~lattis/muon/

I have no idea how to integrate it in CI. The other issue is that analyze will trigger on declare_dependency calls, which are needed when used as a subproject.

**Comment by masatake** [comment]
@neheb, thank you for reporting.
I sent a pull request (#1870) for building fadvise.

**Pull request merged**